bootstrap/api/v1alpha1/rke2config_types.go

```diff
@@ -17,48 +17,314 @@ limitations under the License.
+	// DataSecretName is the name of the secret that stores the bootstrap data script.
+	//+optional
+	DataSecretName *string `json:"dataSecretName,omitempty"`
+
+	// FailureReason will be set on non-retryable errors.
+	//+optional
+	FailureReason string `json:"failureReason,omitempty"`
+
+	// FailureMessage will be set on non-retryable errors.
+	//+optional
+	FailureMessage string `json:"failureMessage,omitempty"`
+
+	// ObservedGeneration is the latest generation observed by the controller.
+	//+optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Conditions defines current service state of the RKE2Config.
+	//+optional
+	Conditions clusterv1.Conditions `json:"conditions,omitempty"`
 }
 
-//+kubebuilder:object:root=true
-//+kubebuilder:subresource:status
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
 
-// Rke2Config is the Schema for the rke2configs API
-type Rke2Config struct {
+// RKE2Config is the Schema for the rke2configs API.
+type RKE2Config struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
-	Spec   Rke2ConfigSpec   `json:"spec,omitempty"`
-	Status Rke2ConfigStatus `json:"status,omitempty"`
+	Spec   RKE2ConfigSpec   `json:"spec,omitempty"`
+	Status RKE2ConfigStatus `json:"status,omitempty"`
+}
+
+func (c *RKE2Config) GetConditions() clusterv1.Conditions {
+	return c.Status.Conditions
+}
+
+func (c *RKE2Config) SetConditions(conditions clusterv1.Conditions) {
+	c.Status.Conditions = conditions
 }
 
-//+kubebuilder:object:root=true
+// +kubebuilder:object:root=true
 
-// Rke2ConfigList contains a list of Rke2Config
-type Rke2ConfigList struct {
+// RKE2ConfigList contains a list of RKE2Config.
+type RKE2ConfigList struct {
 	metav1.TypeMeta `json:",inline"`
 	metav1.ListMeta `json:"metadata,omitempty"`
-	Items           []Rke2Config `json:"items"`
+	Items           []RKE2Config `json:"items"`
+}
+
+// CISProfile defines the CIS Benchmark profile to be activated in RKE2.
+type CISProfile string
+
+const (
+	// CIS1_23 references RKE2's CIS Profile "cis-1.23"
+	CIS1_23 CISProfile = "cis-1.23"
+)
+
+// Encoding specifies the cloud-init file encoding.
+type Encoding string
+
+const (
+	// Base64 implies the contents of the file are encoded as base64.
+	Base64 Encoding = "base64"
+	// Gzip implies the contents of the file are encoded with gzip.
+	Gzip Encoding = "gzip"
+	// GzipBase64 implies the contents of the file are first base64 encoded and then gzip encoded.
+	GzipBase64 Encoding = "gzip+base64"
+)
+
+// File defines the input for generating write_files in cloud-init.
+type File struct {
+	// Path specifies the full path on disk where to store the file.
+	Path string `json:"path"`
+
+	// Owner specifies the ownership of the file, e.g. "root:root".
+	//+optional
+	Owner string `json:"owner,omitempty"`
+
+	// Permissions specifies the permissions to assign to the file, e.g. "0640".
+	//+optional
+	Permissions string `json:"permissions,omitempty"`
```

Comment:
What format of string is expected here? Could we add a validation regex?

Reply:
Yes, definitely.

This part comes BTW from the `Kthrees` provider, which creates a `write_files` primitive in Cloud-init. That is documented [here](https://cloudinit.readthedocs.io/en/latest/topics/modules.html?highlight=write_files#write-files).

Basically, it is an octal string for file permissions, either `'0644' or '0o644'` for a file that is read-write for the owner, read-only for all others, without execution rights. Validation would probably be as such:
- Length can be 4 or 5
- First character should be `0`
- If Length is 4, then each character could be any digit between `0` and `7`.
- If Length is 5, then second character should be 'o' and all other characters should be any digit between `0` and `7`.

So something like this:
`0(|o)[0-7]{3}`